### ISSUE TITLE
Java test breaks in IntelliJ, builds w/ Bazel due to classpath issue

### ISSUE THREAD

**Issue by ulfjack** [bazelbuild/intellij#1641]
I had a fun case earlier today where a normal java_test rule broke in IntelliJ with an error message that it was unable to find assertThrows. I tracked this down to a classpath issue, namely, Bazel puts its test runner on the classpath for java_test rules, which contains an old copy of junit, and somehow, this copy ended up on the classpath before my newer copy of junit which is in the deps of my java_test rule.

As a workaround, I enabled ```--explicit_java_test_deps```, which makes Bazel not put that jar on the classpath and - apparently - makes IntelliJ not pick it up. Ideally, upstream would enable that flag by default. I'm mostly filing this to document the issue.

**Comment by brendandouglas** [comment]
Was this just a standard 'blaze test' invocation, or was it debugging, or fast run/debug?

We could certainly look at enabling the flag on the plugin side by default, though the time is probably better spent pushing for the bazel-side change.

**Comment by ulfjack** [comment]
IntelliJ showed it as a red squiggly line in the editor, Bazel test worked fine.

**Comment by brendandouglas** [comment]
Ah, so the test ran fine, but there's a spurious error in the IDE? Could you link to a project reproducing this?

**Comment by plaird** [comment]
@ulfjack I agree that Bazel should enable explicit java deps by default. I posted that to bazel-discuss but I didn't feel I got much traction for it:

https://groups.google.com/d/msg/bazel-discuss/COJvRi2W1ok/6xBsNbPwBAAJ

My use case was for the IDE classpath implementation in the Bazel Eclipse Feature. For BEF, we now detect the state of the flag and conditionally add the test runner to the classpath if it is 'false'. Yuck, I don't like it.

https://github.com/salesforce/bazel-eclipse/issues/43
